File: group 4/variant12/EventRegistrationService.java
public final class EventRegistrationService {
    public record Request(int participants, int daysBefore, boolean studentClaimed, boolean allowWaitlist) {}

    public record Result(String status, int participants, long total) {}

    public interface EventCapacity {
        int freeSeats();
    }

    public interface StudentRegistry {
        boolean isEligible();
    }

    public interface RegistrationRepository {
        void save(Request request, Result result);
    }

    private final EventCapacity capacity;
    private final StudentRegistry students;
    private final RegistrationRepository registrations;

    public EventRegistrationService(EventCapacity capacity, StudentRegistry students, RegistrationRepository registrations) {
        this.capacity = java.util.Objects.requireNonNull(capacity, "Залежність не задано: capacity");
        this.students = java.util.Objects.requireNonNull(students, "Залежність не задано: students");
        this.registrations = java.util.Objects.requireNonNull(registrations, "Залежність не задано: registrations");
    }

    public Result process(Request request) {
        if (request == null) {
            throw new IllegalArgumentException("Заявку не задано");
        }
        checkRange(request.participants(), 1, 6, "Кількість учасників");
        checkRange(request.daysBefore(), 0, 179, "Дні до події");
        if (request.daysBefore() == 0) {
            return rejected("Реєстрацію закрито");
        }
        int seats = capacity.freeSeats();
        if (seats < 0) {
            throw new IllegalStateException("Від’ємна кількість вільних місць");
        }
        if (seats < request.participants()) {
            if (!request.allowWaitlist()) {
                return rejected("Немає місць");
            }
            Result result = new Result("У списку очікування", request.participants(), 0);
            registrations.save(request, result);
            return result;
        }
        boolean student = request.studentClaimed() && students.isEligible();
        long base = request.participants() * 30_000L;
        int discount = student ? 20 : 0;
        long total = base - base * discount / 100;
        if (request.daysBefore() > 30) {
            total -= 2_000;
        }
        Result result = new Result("Зареєстровано", request.participants(), total);
        registrations.save(request, result);
        return result;
    }

    private static Result rejected(String status) {
        return new Result(status, 0, 0);
    }

    private static void checkRange(int value, int minimum, int maximum, String field) {
        if (value < minimum || value > maximum) {
            throw new IllegalArgumentException("Недопустиме значення: " + field);
        }
    }
}

File: group 3/variant12/event_registration_service.py
from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True)
class Request:
    participants: int
    days_before: int
    student_claimed: bool
    allow_waitlist: bool


@dataclass(frozen=True)
class Result:
    status: str
    participants: int = 0
    total: int = 0


class EventCapacity(Protocol):
    def free_seats(self) -> int:
        ...

class StudentRegistry(Protocol):
    def is_eligible(self) -> bool:
        ...

class RegistrationRepository(Protocol):
    def save(self, request: Request, result: Result) -> None:
        """Фіксує всю операцію або повідомляє про помилку без часткових змін."""
        ...


def _range(value: int, minimum: int, maximum: int, field: str) -> None:
    if type(value) is not int or not minimum <= value <= maximum:
        raise ValueError("Недопустиме значення: " + field)


def _count(value: int, field: str) -> int:
    if type(value) is not int or value < 0:
        raise RuntimeError("Некоректні відомості: " + field)
    return value


def _flag(value: bool, field: str) -> bool:
    if type(value) is not bool:
        raise RuntimeError("Некоректні відомості: " + field)
    return value


class EventRegistrationService:
    def __init__(self, capacity: EventCapacity, students: StudentRegistry, registrations: RegistrationRepository) -> None:
        if capacity is None or students is None or registrations is None:
            raise ValueError("Усі залежності мають бути задані")
        self._capacity = capacity
        self._students = students
        self._registrations = registrations

    def process(self, request: Request) -> Result:
        if not isinstance(request, Request):
            raise ValueError("Заявку не задано або її формат некоректний")
        _range(request.participants, 1, 6, "Кількість учасників")
        _range(request.days_before, 0, 179, "Дні до події")
        if type(request.student_claimed) is not bool:
            raise ValueError("Недопустима логічна ознака: student_claimed")
        if type(request.allow_waitlist) is not bool:
            raise ValueError("Недопустима логічна ознака: allow_waitlist")
        if request.days_before == 0:
            return Result("Реєстрацію закрито")
        seats = _count(self._capacity.free_seats(), "Вільні місця")
        if seats < request.participants:
            if not request.allow_waitlist:
                return Result("Немає місць")
            result = Result("У списку очікування", request.participants, 0)
            self._registrations.save(request, result)
            return result
        student = request.student_claimed and _flag(self._students.is_eligible(), "Студентська пільга")
        base = request.participants * 30_000
        discount = 20 if student else 0
        total = base - base * discount // 100
        if request.days_before > 30:
            total -= 2_000
        result = Result("Зареєстровано", request.participants, total)
        self._registrations.save(request, result)
        return result
